ts/session/apis/open_group_api/sogsv3/sogsV3MutationCache.ts

```diff
@@ -0,0 +1,137 @@
+}
+
+type ReactionAction = 'ADD' | 'REMOVE' | 'CLEAR';
+
+type ReactionChange = {
+  messageId: number; // will be serverId of the reacted message
+  emoji: string;
+  action: ReactionAction;
+};
+
+export type SogsV3Mutation = {
+  seqno: number | null; // null until mutating API request returns
+  server: string; // serverUrl
+  room: string; // roomId
+  changeType: ChangeType;
+  metadata: ReactionChange; // For now we only support message reactions
+};
+
+// we don't want to export this, we want to export functions that manipulate it
+const sogsMutationCache: Array<SogsV3Mutation> = [];
+
+function verifyEntry(entry: SogsV3Mutation): boolean {
+  return Boolean(
+    !entry.server ||
+      !entry.room ||
+      entry.seqno !== null ||
+      entry.metadata.messageId ||
+      entry.metadata.emoji ||
+      entry.metadata.action === 'ADD' ||
+      entry.metadata.action === 'REMOVE'
+  );
+}
+
+export function addToMutationCache(entry: SogsV3Mutation) {
+  if (!verifyEntry(entry)) {
+    window.log.error('SOGS Mutation Cache: Entry verification on add failed!', entry);
+  } else {
+    sogsMutationCache.push(entry);
+    window.log.info('SOGS Mutation Cache: Entry added!', entry);
+  }
+}
+
+export function updateMutationCache(entry: SogsV3Mutation, seqno: number) {
+  if (!verifyEntry(entry)) {
+    window.log.error('SOGS Mutation Cache: Entry verification on update failed!');
+  } else {
+    const entryIndex = findIndex(sogsMutationCache, entry);
+    if (entryIndex >= 0) {
+      const updatedEntry = entry;
+      updatedEntry.seqno = seqno;
+      sogsMutationCache[entryIndex] = updatedEntry;
+      window.log.info('SOGS Mutation Cache: Entry updated!', updatedEntry);
+    } else {
+      window.log.error('SOGS Mutation Cache: Updated failed! Cannot find entry', entry);
+    }
+  }
+}
+
+export async function processMessagesUsingCache(
+  server: string,
+  room: string,
+  message: OpenGroupMessageV4
+) {
+  const updatedReactions = message.reactions;
+  const roomMatches: Array<SogsV3Mutation> = filter(sogsMutationCache, { server, room });
+
+  if (roomMatches?.length) {
+    for (const roomMatch of roomMatches) {
+      if (message.seqno && roomMatch.seqno && roomMatch.seqno <= message.seqno) {
+        const removedEntry = remove(sogsMutationCache, roomMatch);
+        window.log.info('SOGS Mutation Cache: Entry ignored and removed!', removedEntry);
+      } else if (
+        !message.seqno ||
+        (message.seqno && roomMatch.seqno && roomMatch.seqno > message.seqno)
+      ) {
+        for (const reaction of Object.keys(message.reactions)) {
+          const reactionMatches = filter(sogsMutationCache, {
+            server,
+            room,
+            changeType: ChangeType.REACTIONS,
+            metadata: {
+              messageId: message.id,
+              emoji: reaction,
+            },
+          });
+          if (reactionMatches?.length) {
+            for (const reactionMatch of reactionMatches) {
```

Comment:
same coment a s above. the `if` is useless I suppose

Reply:
Fixed